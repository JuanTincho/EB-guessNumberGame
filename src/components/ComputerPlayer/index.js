import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  Button, Grid, Paper, Typography, withStyles,
} from '@material-ui/core';

const styles = theme => ({
  button: {
    margin: theme.spacing.unit,
    textTransform: 'none',
  },
  paper: {
    padding: theme.spacing.unit * 2,
    textAlign: 'center',
    color: theme.palette.text.secondary,
  },
  textField: {
    marginLeft: theme.spacing.unit,
    marginRight: theme.spacing.unit,
  },
});

class ComputerPlayer extends Component {
  state = {
    minNumber: 0,
    maxNumber: 100,
    guessNumber: -1,
    winState: false,
    gameStarted: false,
  };

  startGame = () => {
    const { minNumber, maxNumber } = this.state;
    this.setState({
      gameStarted: true,
      guessNumber: this.getGuessNumber(minNumber, maxNumber),
    });
  };

  restartGame = () => {
    this.setState({
      gameStarted: false,
      winState: false,
      minNumber: 0,
      maxNumber: 100,
    });
  };

  setGuessNumber = (min, max) => {
    this.setState({
      minNumber: min,
      maxNumber: max,
      guessNumber: this.getGuessNumber(min, max),
    });
  };

  winGame = () => {
    this.setState({
      winState: true,
    });
  };

  getGuessNumber = (min, max) => Math.floor((max + min) / 2);

  render() {
    const { classes } = this.props;
    const {
      gameStarted, guessNumber, winState, minNumber, maxNumber,
    } = this.state;
    return (
      <>
        {!gameStarted ? (
          <>
            <Grid item xs={3}>
              <Paper className={classes.paper}>Think of a number between 1 and 100</Paper>
            </Grid>
            <Grid container justify="center">
              <Button
                variant="contained"
                color="secondary"
                className={classes.button}
                onClick={this.startGame}
              >
                Start game!
              </Button>
            </Grid>
          </>
        ) : (
          <>
            <Grid container justify="center">
              <Grid item xs={3}>
                <Typography variant="h6" className={classes.paper} style={{ textAlign: 'center' }}>
                  {!winState ? `Is your number ${guessNumber}?` : 'Yay!!'}
                </Typography>
              </Grid>
            </Grid>
            {!winState ? (
              <>
                <Grid container justify="center">
                  <Button
                    variant="contained"
                    color="primary"
                    className={classes.button}
                    onClick={() => this.setGuessNumber(minNumber, guessNumber)}
                  >
                    My number is SMALLER
                  </Button>
                  <Button
                    variant="contained"
                    color="primary"
                    className={classes.button}
                    onClick={() => this.setGuessNumber(guessNumber, maxNumber)}
                  >
                    My number is BIGGER
                  </Button>
                </Grid>
                <Grid container justify="center">
                  <Button
                    variant="contained"
                    color="secondary"
                    className={classes.button}
                    onClick={this.winGame}
                  >
                    You guessed!
                  </Button>
                </Grid>
              </>
            ) : (
              <>
                <Grid container justify="center">
                  <Button
                    variant="contained"
                    color="secondary"
                    className={classes.button}
                    onClick={this.restartGame}
                  >
                    Play again!
                  </Button>
                </Grid>
              </>
            )}
          </>
        )}
      </>
    );
  }
}

ComputerPlayer.propTypes = {};

export default withStyles(styles)(ComputerPlayer);
