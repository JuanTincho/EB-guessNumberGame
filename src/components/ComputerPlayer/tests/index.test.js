import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import ComputerPlayer from '../index';

describe('<ComputerPlayer /> tests', () => {
  let component;
  let componentInstance;
  beforeEach(() => {
    component = shallow(<ComputerPlayer />).dive();
    componentInstance = component.instance();
  });

  it('ComputerPlayer shoud match snapshot', () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('start game ', () => {
    expect(componentInstance.state.gameStarted).toBeFalsy();
    componentInstance.getGuessNumber = jest.fn(() => 1);
    componentInstance.startGame();
    expect(componentInstance.state.gameStarted).toBeTruthy();
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Win state', () => {
    expect(componentInstance.state.gameStarted).toBeFalsy();
    expect(componentInstance.state.winState).toBeFalsy();
    componentInstance.startGame();
    componentInstance.winGame();
    expect(componentInstance.state.gameStarted).toBeTruthy();
    expect(componentInstance.state.winState).toBeTruthy();
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('setGuessNumber', () => {
    const mockedMinNumber = 1;
    const mockedMaxNumber = 10;
    componentInstance.getGuessNumber = jest.fn();
    componentInstance.setGuessNumber(mockedMinNumber, mockedMaxNumber);
    expect(componentInstance.state.minNumber).toEqual(mockedMinNumber);
    expect(componentInstance.state.maxNumber).toEqual(mockedMaxNumber);
    expect(componentInstance.getGuessNumber).toHaveBeenCalledTimes(1);
  });

  it('guetGuessNumber', () => {
    expect(componentInstance.getGuessNumber(1, 100)).toBe(50);
    expect(componentInstance.getGuessNumber(50, 100)).toBe(75);
    expect(componentInstance.getGuessNumber(1, 50)).toBe(25);
    expect(componentInstance.getGuessNumber(4, 9)).toBe(6);
  });

  it('restartGame', () => {
    componentInstance.setState({
      gameStarted: true,
      winState: true,
      minNumber: 12,
      maxNumber: 125,
    });
    componentInstance.restartGame();
    expect(componentInstance.state.gameStarted).toBeFalsy();
    expect(componentInstance.state.winState).toBeFalsy();
    expect(componentInstance.state.minNumber).toEqual(0);
    expect(componentInstance.state.maxNumber).toEqual(100);
  });
});
