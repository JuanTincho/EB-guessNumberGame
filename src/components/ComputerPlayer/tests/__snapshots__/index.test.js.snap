// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ComputerPlayer /> tests ComputerPlayer shoud match snapshot 1`] = `
Array [
  <div
    className="MuiGrid-item-5 MuiGrid-grid-xs-3-35"
  >
    <div
      className="MuiPaper-root-101 MuiPaper-elevation2-105 MuiPaper-rounded-102 ComputerPlayer-paper-2"
    >
      Think of a number between 1 and 100
    </div>
  </div>,
  <div
    className="MuiGrid-container-4 MuiGrid-justify-xs-center-21"
  >
    <button
      className="MuiButtonBase-root-154 MuiButton-root-128 MuiButton-contained-139 MuiButton-containedSecondary-141 MuiButton-raised-142 MuiButton-raisedSecondary-144 ComputerPlayer-button-1"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-129"
      >
        Start game!
      </span>
      <span
        className="MuiTouchRipple-root-157"
      />
    </button>
  </div>,
]
`;

exports[`<ComputerPlayer /> tests Win state 1`] = `
Array [
  <div
    className="MuiGrid-item-5 MuiGrid-grid-xs-3-35"
  >
    <div
      className="MuiPaper-root-101 MuiPaper-elevation2-105 MuiPaper-rounded-102 ComputerPlayer-paper-2"
    >
      Think of a number between 1 and 100
    </div>
  </div>,
  <div
    className="MuiGrid-container-4 MuiGrid-justify-xs-center-21"
  >
    <button
      className="MuiButtonBase-root-154 MuiButton-root-128 MuiButton-contained-139 MuiButton-containedSecondary-141 MuiButton-raised-142 MuiButton-raisedSecondary-144 ComputerPlayer-button-1"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-129"
      >
        Start game!
      </span>
      <span
        className="MuiTouchRipple-root-157"
      />
    </button>
  </div>,
]
`;

exports[`<ComputerPlayer /> tests start game  1`] = `
Array [
  <div
    className="MuiGrid-item-5 MuiGrid-grid-xs-3-35"
  >
    <div
      className="MuiPaper-root-101 MuiPaper-elevation2-105 MuiPaper-rounded-102 ComputerPlayer-paper-2"
    >
      Think of a number between 1 and 100
    </div>
  </div>,
  <div
    className="MuiGrid-container-4 MuiGrid-justify-xs-center-21"
  >
    <button
      className="MuiButtonBase-root-154 MuiButton-root-128 MuiButton-contained-139 MuiButton-containedSecondary-141 MuiButton-raised-142 MuiButton-raisedSecondary-144 ComputerPlayer-button-1"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-129"
      >
        Start game!
      </span>
      <span
        className="MuiTouchRipple-root-157"
      />
    </button>
  </div>,
]
`;
